base/compiler/abstractinterpretation.jl

```diff
@@ -23,18 +23,32 @@ end
 
 const empty_bitset = BitSet()
 
+function should_infer_for_effects(sv::InferenceState)
+    sv.ipo_effects.terminates === ALWAYS_TRUE &&
+    sv.ipo_effects.effect_free === ALWAYS_TRUE
+end
+
 function abstract_call_gf_by_type(interp::AbstractInterpreter, @nospecialize(f),
                                   arginfo::ArgInfo, @nospecialize(atype),
                                   sv::InferenceState, max_methods::Int = get_max_methods(sv.mod, interp))
-    if sv.params.unoptimize_throw_blocks && is_stmt_throw_block(get_curr_ssaflag(sv))
+    if !should_infer_for_effects(sv) &&
+            sv.params.unoptimize_throw_blocks &&
+            is_stmt_throw_block(get_curr_ssaflag(sv))
+        # Disable inference of calls in throw blocks, since we're unlikely to
+        # need their types. There is one exception however: If up until now, the
+        # function has not seen any side effects, we would like to make sure there
+        # aren't any in the throw block either to enable other optimizations.
         add_remark!(interp, sv, "Skipped call in throw block")
+        tristate_merge!(sv, Effects(ALWAYS_TRUE, TRISTATE_UNKNOWN,
```

Comment:
No, it's not really used consistently, but `ALWAYS_FALSE` is supposed to indicate that something always happens (e.g. there's always an effect or it always throws). For now it doesn't matter, they're treated the same, but in the future, I may want to be able to refine `TRISTATE_UNKNOWN` into `ALWAYS_TRUE` for a more precise type, but `ALWAYS_FALSE`, should never be able to be refined into `ALWAYS_TRUE`.

Reply:
being ALWAYS_TRUE seems like something this design won't be able to compute, since it is dataflow-sensitive, IIUC

It seems that below you are setting `ALWAYS_TRUE` to mean `MIGHT_BE_TRUE`?

Reply:
Once you're in this block, you're guarantee to throw eventually, which is all that's required for `:consistent` (assuming the rest of the function is). In general, the RHS of a tristate_merge! has the effects that you can conclude from the statement you're currently looking at.